### PR TITLE
[charlie] proposed enhancement: api-server incoming deploy rate limiting

### PR DESCRIPTION
**Pull request by marc-casperlabs**
Untested (but compiling :)) fix that introduces rate limiting for incoming deploys.

Based on a ticketing system, numbers are tweakable through the configuration.

Someone who has been on the front lines should give this a whirl locally.

### CODE REVIEW

**Comment by goral09** [comment]
The high rate of incoming deploys was not a problem but the underlying LMDB issue. Maybe let's first test if the other PR helps before we merge this.

**Review by Fraser999 — APPROVED** [review]

**Review by goral09 — CHANGES_REQUESTED** [review]
Let's really not merge "untested but compiling" fixes.

**Comment by marc-casperlabs** [comment]
> Let's really not merge "untested but compiling" fixes.

@momipsl is currently putting it through the ringer.

**Comment by marc-casperlabs** [comment]
> The high rate of incoming deploys was not a problem but the underlying LMDB issue. Maybe let's first test if the other PR helps before we merge this.

I thought these were separate issues? Incoming deploys being one, while syncing up being another?

**Comment by Fraser999** [comment]
> Let's really not merge "untested but compiling" fixes.

Yeah - I agree that that's not a good idea (although I do appreciate the honesty of the admission in the PR comment!)

For what it's worth, I did test this last night and saw the node rejecting then re-accepting Deploys when I repeatedly sent a Deploy.  I'm not sure the chosen number for the ticketer will be low enough, but we have a lot of unknowns in our config values.

**Comment by goral09** [comment]
> > The high rate of incoming deploys was not a problem but the underlying LMDB issue. Maybe let's first test if the other PR helps before we merge this.
> 
> I thought these were separate issues? Incoming deploys being one, while syncing up being another?

No, I think it was the same root cause – `lmdb_chainspec_store` panicking when not able to get read-only txn lock.

**Comment by marc-casperlabs** [comment]
> No, I think it was the same root cause – lmdb_chainspec_store panicking when not able to get read-only txn lock.

Right. The issue is that we have no backpressure. In the case of "catching up", there's a limit at any given time on how many requests we can ultimately be asked to handle.

However if people are stress-testing the deployment interface, this is essentially unbounded, so we need both, IMHO.

**Comment by EdHastingsCasperLabs** [comment]
> For what it's worth, I did test this last night and saw the node rejecting then re-accepting Deploys when I repeatedly sent a Deploy. I'm not sure the chosen number for the ticketer will be low enough, but we have a lot of unknowns in our config values.

I did half a dozen runs locally with the limits set to 1 and 10 to enable quicker observation and it was promising. We asked @momipsl to give it a whirl with pressure applied and let us know how it goes. I'm waiting on his feedback before making any decisions about this.

**Comment by goral09** [comment]
I don't consider this to be a proper fix, even worse b/c it is leaking to the client. We should try caching the chainspec (instead of returning it on every new deploy) and see if it helps.

**Comment by EdHastingsCasperLabs** [comment]
@momipsl has produced some findings of this branch...https://github.com/CasperLabs/stests-reports/blob/master/0001/report.md

TL;DR: this branch's throttling appears to be more stable and overall more performant.

**Comment by EdHastingsCasperLabs** [comment]
> I don't consider this to be a proper fix, even worse b/c it is leaking to the client. 

I don't agree that it is "_leaking_ to the client"; as per http it is a client error to exceed rate limiting and 429 is the correct http status. https://httpstatuses.com/429

We can't accept unlimited deploys per second; there is a practical limit. Michael specified **100** deploys per second in his [tolerances design document](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/151453943/Network+Operating+Conditions+and+Tolerances#NetworkOperatingConditionsandTolerances-ClientMessageHandling(Deploys)), and Mark's configuration (possibly coincidentally) matches that. 

> We should try caching the chainspec (instead of returning it on every new deploy) and see if it helps.

I agree with this, but think that it is an orthogonal consideration to deploy rate limiting.

**Review by EdHastingsCasperLabs — APPROVED** [review]
The api-server component has received a significant amount of refactoring and enhancement in the master branch, so it would require some amount of factoring to integrate a similar solution in master if we decided to adopt this approach to rate limiting incoming deploys.

However, based upon the promising stest output, I do think this is worth giving a trial run if / when we roll another update of charlie testnet out.

**Comment by goral09** [comment]
OK, let me try to rephrase it:

In general, I agree that we should not accept unlimited deploys and rate limit the client's input. You're also completely right about that there even an HTTP error code for that. My point was that we're not fixing the root cause with this and our node can be much more performant without this. Eventually, we should of course rate limit but by rate limiting right now, we are not giving ourselves a chance to measure the maximum performance of the node (who knows, maybe with chainspec cached it will be able to take on a 1000 or 10000 deploys per second).

**Comment by EdHastingsCasperLabs** [comment]
I acknowledge your points @goral09 ; I think the correct course of action is to wait until @momipsl has had a chance to do a test run using chainspec cache PR and compare the results of this and that before picking a go forward approach.

**Comment by goral09** [comment]
@EdHastingsCasperLabs Of course, that's why both of these PRs are just drafts.

**Comment by marc-casperlabs** [comment]
That's a good point, @goral09. Maybe we should set the default to some ridiculously high number or simply disabled? That way, we can make a distinction between stability in a longer term test net and short term testing.

**Comment by goral09** [comment]
If this works as expected then we could merge it and leave it disabled for now. I guess it is a desired feature anyway. I am not sure whether we need this feature:
> // Allow exceeding the deploy rate for short bursts following periods of inactivity.

Or maybe I just don't understand it yet.

**Comment by marc-casperlabs** [comment]
@goral09 

The burst rate allows exceeding the maximum rate for a set amount of time to fill up buffers.

Image this: You have a max rate of 10 deploys per second set, and send them for 5 seconds. Your client wants to send 100 deploys in total. After 5 seconds there's a hiccup, for one second during which you don't sent any deploys. Finally you send the remaining 50 over the course of another 5 seconds, so all in all 100 deploys were transferred in **11 seconds**.

With a burst limit of, say 30, the same scenario would go like this: First 5 seconds at 10 deploys send the first 50. Now during one second of idleness, the node accumulates 10 "credits" (up to a maximum of 30), so when the client starts sending again, it can exceed the rate of 10 deploys/second until it has "caught up". It may send 20 deploys for one second, following by 4 seconds of 10 deploys per second, or even 12 deploys per second for 5 seconds.

As a result, 100 deploys will be transferred in **10 seconds**.

**Comment by goral09** [comment]
Thanks for explaining Marc. I don't think this approach will work though – our problem is that there's a limit for the number of events that an incoming deploy creates. Even a momentary burst that exceeds the limit can kill the node.

**Comment by marc-casperlabs** [comment]
Closing this for now, as it will not be in charlie. Future testnets will want this feature (see story).
